### ISSUE TITLE
Is there a paradigm where I can use a "controlled component" in livewire

### ISSUE THREAD

**Issue by roni-estein** [livewire/livewire#346]
I'm working on my second livewire charity project... I need to know it better before I introduce it into something paid. I'm hoping I can make that choice soon.

My first project was a tiny set of forms and everything mapped 1:1 where data was either shown or not shown, and I'd like to have a series of slightly more complicated controls.

I have a series of yes no questions where yes or no, depending on the question will reveal a text are for further inquiry. I'd like to bind that data to my parent form.

Here is a visual:

<img width="343" alt="Screen Shot 2019-10-24 at 1 06 04 PM" src="https://user-images.githubusercontent.com/8517475/67521823-3ad0a380-f671-11e9-92e6-fd5e4c1db3e8.png">

This is the expanded version.

<img width="351" alt="Screen Shot 2019-10-24 at 1 06 27 PM" src="https://user-images.githubusercontent.com/8517475/67521836-415f1b00-f671-11e9-8e9a-1f13b8a36dd8.png">

I'm trying to think of the right paradigm to map these yes no components into the larger form.

Here is a look at the component code.

```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class QuestionWithHiddenOther extends Component
{
    
    
    public $hideOther;
    public $hideOn = 1;
    public $name = '';
    public $label = '';
    public $question = '';
    public $answer = '';
    
    public function mount($name, $question, $label ='',$hideOn=1)
    {
        $this->name = $name;
        $this->question = $question;
        $this->label = $label;
        $this->hideOn = $hideOn;
        $this->hideOther = $hideOn;
    }
    public function render()
    {
        return view('livewire.question-with-hidden-other');
    }
    
    public function updated()
    {
        $this->emit('updatedQuestionWithOther', $this->name, $this->hideOn, $this->hideOther, $this->answer);
    }
}
```

here is the blade

```php
<div class="py-1">
    <div class="flex flex-col">
        <div class="flex items-center">
            <div class="w-2/3">{{ $question }}</div>
            <div class="ml-2 w-1/3 text-right">
                <label class="inline-flex items-center">
                    <input type="radio" wire:model.lazy="hideOther" name="{{ $name }}_yes" id="{{ $name }}_yes" class="form-radio" value="1" @checked($hideOther,1)>
                    <span class="ml-1 mt-1">Yes</span>
                </label>
                <label class="ml-2 inline-flex items-center">
                    <input type="radio" wire:model.lazy="hideOther" name="{{ $name }}_no" id="{{ $name }}_no" class="form-radio" value="0" @checked($hideOther,0)>
                    <span class="ml-1 mt-1">No</span>
                </label>
            </div>
        </div>

        @if($hideOther != $hideOn)
        <label class="block mt-2 mb-6" wire:transition.slide.up.500ms>
            <div class="form-label"> {{ $label }}: @error('answer')<br>{{ $message }}@enderror</div>
            <textarea wire:model.lazy="answer"
                      name="answer"
                      id="answer"
                      autocomplete="" autocapitalize="off" rows="5"
                class="form-input block w-full @error('answer') error @enderror">
            </textarea>
        </label>
        @endif

    </div>
</div>
```

I'm calling it from inside my form component as follows
```
 @livewire('question-with-hidden-other','well_today', 'Are you well today?','If no, describe')

    @livewire('question-with-hidden-other','allergies', 'Do you have any allergies?','If yes, describe',0)

    @livewire('question-with-hidden-other','conditions_following_vaccines', 'Have you ever had a serious reaction or condition following any vaccine?','If yes, describe',0)
    @livewire('question-with-hidden-other','conditions_requiring_dr', 'Do you have any conditions that require regular visits to a doctor?','If yes, describe and discuss with immunizer',0)
```

The question is what a good reactive paradigm?

right now, I'm just passing the event up to the main component.

```php
 protected $listeners = ['updatedQuestionWithOther' => 'updateQuestionsWithOther'];
    
    public function updateQuestionsWithOther($name, $hideOn, $hideOther, $answer)
    {
        // do something here
    }
```
Another issue is once validation comes in I'll have to change answers to a keyed component presumably thats what the key() livewire function does, otherwise the error bag will fail, unless I drop the questions into an array.

I'm not sure, this just felt like weak tea. Should I be validating in the small component individually? Is there a way I can wire:model the whole element sub component, like a value prop and input event that wire model will just listen to? If there is, how do I let the parent component know to bind the child component with reactive data?

Thanks for any advice you have. Am I looking at this too much like vue?
Cheers,

**Comment by roni-estein** [comment]
Update, even worse :

I thought I'd clean it up a bit:

in component added:
```php

in the component public $otherQuestions = [
        'well_today' => [
            'text' => 'Are you well today?',
            'label' => 'If no, describe',
            'hide_on' => 1,
        ],
        'allergies' => [
            'text' => 'Do you have any allergies?',
            'label' => 'If yes, describe',
            'hide_on' => 0,
        ],
        'conditions_following_vaccines' => [
            'text' => 'Have you ever had a serious reaction or condition following any vaccine?',
            'label' => 'If yes, describe',
            'hide_on' => 0,
        ],
        'conditions_requiring_dr' => [
            'text' => 'Do you have any conditions that require regular visits to a doctor?',
            'label' => 'If yes, describe and discuss with immunizer',
            'hide_on' => 0,
        ],
    ];
```

Then in blade refactored:
```
@foreach($otherQuestions as $name => $question)
{{--        @dump([$name,$question])--}}
        @livewire('question-with-hidden-other',$name, $question['text'],$question['label'], $question['hide_on'])

@endforeach
```

Looks the same but blows but only after it renders properly and you toggle some changes. Which all used to work until organizing data in an array for easier upgrades.

Here are the images:
First Render:
<img width="369" alt="Screen Shot 2019-10-24 at 4 27 14 PM" src="https://user-images.githubusercontent.com/8517475/67526885-52ad2500-f67b-11e9-9b18-28fd46607432.png">

After switching from yes to no on top item:
<img width="393" alt="Screen Shot 2019-10-24 at 4 27 24 PM" src="https://user-images.githubusercontent.com/8517475/67526919-6062aa80-f67b-11e9-81f1-b0b5579fc405.png">

then re-hitting either radio element

ErrorException
hash_equals(): Expected user_string to be a string, null given
http://localhost:3000/livewire/message/null

**Comment by roni-estein** [comment]
I could put a demo together if it helps debug it later tonight.

**Comment by roni-estein** [comment]
Weird, changing this back to:

@livewire('question-with-hidden-other','well_today', 'Are you well today?','If no, describe')
    @livewire('question-with-hidden-other','allergies', 'Do you have any allergies?','If yes, describe',0)
    @livewire('question-with-hidden-other','conditions_following_vaccines', 'Have you ever had a serious reaction or condition following any vaccine?','If yes, describe',0)
    @livewire('question-with-hidden-other','conditions_requiring_dr', 'Do you have any conditions that require regular visits to a doctor?','If yes, describe and discuss with immunizer',0)

And everything works again?!? I know I'm missing something, but I can't use something like this is production products as much as I want to. At least with vue, even though it's extra work and less testable, I always know whats going to happen. But I'll try and grok this more when I have spare time. I'll throw together an example project shortly so it can be hacked on. 

Lastly, if someone knows how to make this more of a controlled component, to clarify. So in a perfect world. I'd like to wire:model a component that I make,  so in the component I'd like to have some way to pass in a value and get back input events automagically. That would be great.

**Comment by roni-estein** [comment]
Here is a project with instructions and the bug reproduced.

https://github.com/roni-estein/livewire-bug

**Comment by roni-estein** [comment]
I'm a douche! I didn't see the keyed component section!
